overlord/assertstate/assertstate.go

```diff
@@ -773,11 +773,127 @@ func validationSetAssertionForEnforce(st *state.State, accountID, name string, s
+		accountID, name, sequence, err := snapasserts.ParseValidationSet(vsstr)
+		if err != nil {
+			return err
+		}
+
+		// try to get existing from the db
+		headers := map[string]string{
+			"series":     release.Series,
+			"account-id": accountID,
+			"name":       name,
+		}
+		if sequence > 0 {
+			headers["sequence"] = fmt.Sprintf("%d", sequence)
+		}
+		atSeq := &asserts.AtSequence{
+			Type:        asserts.ValidationSetType,
+			SequenceKey: []string{release.Series, accountID, name},
+			Sequence:    sequence,
+			Revision:    asserts.RevisionNotKnown,
+			Pinned:      sequence > 0,
+		}
+
+		// prepare tracking data, note current is not known yet
+		tr := &ValidationSetTracking{
+			AccountID: headers["account-id"],
+			Name:      headers["name"],
+			Mode:      Enforce,
+			// may be 0 meaning no pinning
+			PinnedAt: sequence,
+		}
+
+		extraVsHeaders = append(extraVsHeaders, headers)
+		newTracking = append(newTracking, tr)
+
+		vs, err := getSpecificSequenceOrLatest(db, headers)
+		// found locally
+		if err == nil {
+			// update with pool
+			atSeq.Sequence = vs.Sequence()
+			atSeq.Revision = vs.Revision()
+			if err := pool.AddSequenceToUpdate(atSeq, atSeq.Unique()); err != nil {
+				return err
+			}
+		} else {
+			if !asserts.IsNotFound(err) {
+				return err
+			}
+			// try to resolve with pool
+			if err := pool.AddUnresolvedSequence(atSeq, atSeq.Unique()); err != nil {
+				return err
+			}
+		}
+	}
+
+	checkBeforeCommit := func(db *asserts.Database, bs asserts.Backstore) error {
+		tmpDb := db.WithStackedBackstore(bs)
+		// get the resolved validation set asserts, add to validation sets and check
+		var extraVs []*asserts.ValidationSet
+		for _, headers := range extraVsHeaders {
+			vs, err := getSpecificSequenceOrLatest(tmpDb, headers)
+			if err != nil {
+				return fmt.Errorf("internal error: cannot find validation set assertion: %v", err)
+			}
+			extraVs = append(extraVs, vs)
+		}
+
+		valsets, err := EnforcedValidationSets(st, extraVs...)
+		if err != nil {
+			return err
+		}
+		if err := valsets.Conflict(); err != nil {
+			return err
+		}
+		if err := valsets.CheckInstalledSnaps(snaps, ignoreValidation); err != nil {
+			// the returned error may be ValidationSetsValidationError which is normal and means we cannot enforce
+			// the new validation sets - the caller should resolve the error and retry.
+			return err
+		}
+
+		// all fine, will be committed (along with its prerequisites if any) on
+		// return by resolvePoolNoFallback
+		return nil
+	}
+
+	opts := &RefreshAssertionsOptions{}
+	if err := resolvePoolNoFallback(st, pool, checkBeforeCommit, userID, deviceCtx, opts); err != nil {
+		return err
+	}
+
+	// no error, all validation-sets can be enforced, update tracking for all vsets
+	for i := 0; i < len(extraVsHeaders); i++ {
+		// get latest assertion from the db to determine current
+		a, err := db.FindSequence(asserts.ValidationSetType, extraVsHeaders[i], -1, -1)
```

Comment:
nitpick: looks slightly simpler to me like this but feel free to ignore
```suggestion
	for i, headers := range extraVsHeaders {
		// get latest assertion from the db to determine current
		a, err := db.FindSequence(asserts.ValidationSetType, headers, -1, -1)
```

Reply:
+1

Another question (maybe @pedronis can help): this code calls `db.FindSequence()`, but based on the comment ("get latest assertion from the db to determine current") and the pre-commit validation code above, it would seem to me that we should call `getSpecificSequenceOrLatest()` instead.

Reply:
no, Current tracks the latest local sequence for the validation set, you can see that the corresponding code in EnforceValidationSet goes through validationSetAssertionForEnforce and there uses getLatest for the value that ends up in Current

Reply:
OK. I can't really say that I understand the logic, but I have verified that indeed the TryEnforceValidationSets function logic matches the checks in EnforceValidationSet, so I trust it that it's right.

@MiguelPires I've pushed a commit implementing your suggestion.